### DISCUSSION TITLE
How can I have better types on response classes?

### DISCUSSION THREAD

**Discussion by Johnhersh** [vitalik/django-ninja#829] [Q&A] [answered]
Hi all!

I'm converting our project from Django to Django Ninja. Before we used to return a response object that would look something like this:

```python
class AssetResponse:
    def __init__(self, text, url, thumbnail):
        self.text = text
        self.url = url
        self.thumbnail = thumbnail

    def to_dict(self) -> dict:
        return {
            "name": self.text,
            "url": self.url,
            "thumbnail": self.thumbnail,
        }

    def to_asset(self) -> Asset:
        return Asset(name=self.text, url=self.url)
```

The nice thing here is I can use mypy to make sure the class is created correctly. So nobody can do `response = AssetResponse(Text, URL, ThUmBnail) without seeing an error.

But now I converted it to a schema:
```python
class AssetResponse(Schema):
    text: str
    url: str
    thumbnail: str
```

But when creating this class it shows as:
`class PromptResponse(**data: Any)`

With **data: Any, so I can no longer ensure the types.

Is there any way around this? Or am I structuring this wrong in the first place?

**Comment by vitalik** [comment]
@Johnhersh well this is default behaviour of pydantic.BaseModel (and Schema inherited from it) 

as it can validate all sort of input data and turn it into a valid types

Personally I do not see why you would need to create Schema objects directly as validation of the response object will happen automatically:


```Python

class AssetResponse(Schema):
    text: str
    url: str
    thumbnail: str


@api.get('/asset', response=AssetResponse)
def get_asset(request):
    return {
       "text": "foo",
       "url": "https://gihub.com",
       "thumbnail": "/some.jpg",
    }

```

**Reply by Johnhersh** [reply]
That's fine, but the problem with that is if I do something like this:
```python
class AssetResponse(Schema):
    text: str
    url: str
    thumbnail: str


@api.get("/asset", response=AssetResponse)
def get_asset(request):
    return {
        "TeXt": "foo",
        "uRl": "https://gihub.com",
        "thumbNail": "/some.jpg",
    }
```

There's zero indication by Pylint or Mypy that anything is wrong. Sure we'll find it as soon as someone hits that path and it'll throw errors. But with the regular class with `__init__` Pylint and Mypy will complain before it even gets to that.

**Reply by vitalik** [reply, marked as answer]
yeah, indeed at this moment there are no nice solution

but I'm planning somewhere in the future add suppport for [pydantic dataclasses](https://docs.pydantic.dev/1.10/usage/dataclasses/) that should basically cover your case

**Reply by Johnhersh** [reply]
That would make sense! Looking forward to it 🙂
